File: source_code - v1/FUNC_randomClustering.m
function [MODEL] = FUNC_randomClustering(MODEL,B,t_thres,theta,Area)
numSAT = size(MODEL.SAT,2); numUE = size(MODEL.UE,2);
% Distance from SATs to users, row-SAT, column-UE
Distance = sqrt( (MODEL.SAT(1,:)'-MODEL.UE(1,:)).^2 + ...
                 (MODEL.SAT(2,:)'-MODEL.UE(2,:)).^2 + ...
                 (MODEL.SAT(3,:)'-MODEL.UE(3,:)).^2 );
% Convert from t_thres to d_thres
d_thres = t_thres * 3e8;
% Find cannot-links
C_not = Distance > d_thres; % if C_not(m,u)=1 then m-u is cannot-link
IdClusUE = zeros(1,numUE);
% Initialize randomly all clusters
NumUE_in_SAT = zeros(1,numSAT);
numLoop = 0;
while(1)
    numLoop = numLoop + 1;
    U_free = find(IdClusUE==0);
    if(isempty(U_free))
        break
    end
    for iu = 1:size(U_free,2)
        u = U_free(iu);
        % Consider UEs that have not been served yet
        if(IdClusUE(u)==0)
            % Find a free near SAT and the number of UEs conneting less
            % than B
            NearSAT = find(Distance(:,u)<d_thres);
            Free_SAT = find(NumUE_in_SAT < B);
            [OkSAT,pos]=intersect(NearSAT,Free_SAT);
            OkSAT = OkSAT';
            if(isempty(OkSAT)) % worst case
                % Connect to the nearest SAT in the worst case
                [dismin,ind_SAT] = min(Distance(:,u));
                % Delete a UE in the nearest SAT
                U_in_indSAT = find(IdClusUE==ind_SAT);
                IdClusUE(U_in_indSAT(1)) = 0;
                % The u-th UE is served by (ind_SAT)-th SAT
                IdClusUE(u) = ind_SAT;
            else % random case
                ind_SAT = OkSAT(randi([1,size(OkSAT,2)],1));
                % The number UEs associate with (ind_SAT)-th SAT
                NumUE_in_SAT(ind_SAT) = NumUE_in_SAT(ind_SAT) + 1;
                % The u-th UE is served by (ind_SAT)-th SAT
                IdClusUE(u) = ind_SAT;
            end
        end
    end
    if(numLoop == 1000)
        error("Cannot clustering due to strict constraints");
    end
end
% % show random clustering
% MODEL.IdClusUE = IdClusUE;
% MODEL.NumUE_in_SAT = NumUE_in_SAT;
% FUNC_Plot_model_2D(MODEL, Area, "cluster")

MODEL.IdClusUE = IdClusUE;
MODEL.NumUE_in_SAT = NumUE_in_SAT;
end


%% Mini functions
function value = func_utility(Distance,CHAN,cluster,IdClusUE,theta,T_total,H_total)
% cluster is related to SAT m
IdUE = find(IdClusUE==cluster);
value = 0;
for iu = 1:size(IdUE,2)
    u = IdUE(iu);
    value = value + theta*Distance(cluster,u)/3e8/T_total...
            - (1-theta)*CHAN(cluster,u)/H_total;
end
end

File: source_code - v1/FUNC_randomClustering.py
import numpy as np


def FUNC_randomClustering(MODEL, B, t_thres, theta, Area):
    sat = MODEL["SAT"]
    ue = MODEL["UE"]
    numSAT = sat.shape[1]
    numUE = ue.shape[1]
    distance = np.sqrt(
        (sat[0, :][:, None] - ue[0, :][None, :]) ** 2
        + (sat[1, :][:, None] - ue[1, :][None, :]) ** 2
        + (sat[2, :][:, None] - ue[2, :][None, :]) ** 2
    )
    d_thres = t_thres * 3e8
    id_clus_ue = np.zeros(numUE, dtype=int)
    numUE_in_SAT = np.zeros(numSAT, dtype=int)
    numLoop = 0

    while True:
        numLoop += 1
        u_free = np.where(id_clus_ue == 0)[0]
        if u_free.size == 0:
            break

        for u in u_free:
            if id_clus_ue[u] == 0:
                near_sat = np.where(distance[:, u] < d_thres)[0] + 1
                free_sat = np.where(numUE_in_SAT < B)[0] + 1
                ok_sat = np.intersect1d(near_sat, free_sat)
                if ok_sat.size == 0:
                    ind_sat = int(np.argmin(distance[:, u])) + 1
                    users_in_sat = np.where(id_clus_ue == ind_sat)[0]
                    if users_in_sat.size > 0:
                        id_clus_ue[users_in_sat[0]] = 0
                    id_clus_ue[u] = ind_sat
                else:
                    ind_sat = int(np.random.choice(ok_sat))
                    numUE_in_SAT[ind_sat - 1] += 1
                    id_clus_ue[u] = ind_sat

        if numLoop == 1000:
            raise RuntimeError("Cannot perform clustering due to strict constraints")

    MODEL["IdClusUE"] = id_clus_ue
    MODEL["NumUE_in_SAT"] = numUE_in_SAT
    return MODEL
